File: coding-challenges/sources/leetcode/219. Contains Duplicate II/219-ContainsDuplicateII.kt
fun main() {
    println(containsNearbyDuplicate(intArrayOf(1, 2, 3, 1), 3) == true)
    println(containsNearbyDuplicate(intArrayOf(1, 0, 1, 1), 1)  == true)
    println(containsNearbyDuplicate(intArrayOf(1, 2, 3, 4), 2) == false)
}

fun containsNearbyDuplicate(nums: IntArray, k: Int): Boolean {
    // create an empty set
    var seen = mutableSetOf<Int>()
    // iterate over the elements of the array
    nums.forEachIndexed { i, num ->
        // if the element exists in the set
        if (num in seen) return true
        // add the element to the set
        seen.add(num)
        // if the size of the set exceeds k, remove the oldenst element
        if (i >= k) seen.remove(nums[i - k])
    }
    // no such elements were found
    return false
}

File: coding-challenges/sources/leetcode/219. Contains Duplicate II/219-ContainsDuplicateII.swift


fileprivate func containsNearbyDuplicate(_ nums: [Int], _ k: Int) -> Bool {
    // create an empty set
    var seen = Set<Int>()

    // iterate over the elements of the array
    for (i, num) in nums.enumerated() {
        // if the element exists in the set
        if seen.contains(num) {
            return true
        }

        // add the element to the set
        seen.insert(num)

        // if the size of the set exceeds k, remove the oldenst element
        if i >= k {
            seen.remove(nums[i - k])
        }
    }

    // no such elements were found
    return false
}

import XCTest

final class ContainsDuplicateIITests: XCTestCase {
    func test() {
        XCTAssertTrue(containsNearbyDuplicate([1,2,3,1], 3))
        XCTAssertTrue(containsNearbyDuplicate([1,0,1,1], 1))
        XCTAssertFalse(containsNearbyDuplicate([1,2,3,1,2,3], 2))
    }
}
